signature.js

```diff
@@ -1,35 +1,54 @@
-const CryptoJS = require('crypto-js')
+const crypto = require('crypto')
 
 const base64url = (src) => {
   var srcBuffer = Buffer.from(JSON.stringify(src), 'utf8')
   let encodedSrc = srcBuffer.toString('base64').replace(/=+$/, '')
   return encodedSrc
 }
 
-const sign = (msg) => {
-  const hmac = CryptoJS.HmacSHA256(msg, 'secret-goes-here')
-  return hmac.toString(CryptoJS.enc.Base64).replace(/=+$/, '')
+const tokenHeader = { "alg": "RS256", "typ": "JWT" }
+
+const buildSign = (privateKey) => {
+  return (msg) => {
+    const s = crypto.createSign('SHA256')
+    s.write(msg)
+    s.end()
+    return s.sign(privateKey, 'base64')
+  }
 }
 
-const tokenHeader = { "alg": "HS256", "typ": "JWT" }
+const buildGenerate = (sign) => {
+  return (payload) => {
+    let msg = base64url(tokenHeader) + '.' + base64url(payload)
+    let signature = sign(msg)
+    let token = msg + '.' + signature
+    return token
+  }
+}
 
-const generate = (payload) => {
-  let msg = base64url(tokenHeader) + '.' + base64url(payload)
-  let signature = sign(msg)
-  let token = msg + '.' + signature
-  return token
+const buildVerify = (publicKey) => {
+  return (token) => {
+    let components = token.split('.', 3)
+    if(components.length !== 3) { return false }
+    let msg = components[0] + '.' + components[1]
+    let providedSignature = components[2]
+    let v = crypto.createVerify('SHA256')
+    v.write(msg)
+    v.end()
+    return v.verify(publicKey, providedSignature, 'base64')
+  }
 }
 
-const verify = (token) => {
-  let components = token.split('.', 3)
-  if(components.length !== 3) { return false }
-  let msg = components[0] + '.' + components[1]
-  let providedSignature = components[2]
-  let computedSignature = sign(msg)
-  return providedSignature === computedSignature
+const build = (privateKey, publicKey) => {
+  return {
+    generate: buildGenerate(buildSign(crypto.createPrivateKey(privateKey))),
+    verify: buildVerify(crypto.createPublicKey(publicKey)),
+    publicKey: publicKey
+  }
 }
 
-module.exports = {
-  generate: generate,
-  verify: verify
+module.exports = (options = {}) => {
+  if(options.privateKey && options.publicKey) {
```

Comment:
will there be error handling should the public and private keys not exist?

Reply:
It "handles" the error by not producing the signature service. And then leaves it up to the caller what to do next. We will probably want a more standard convention once we have more than one service (or more than one use of this service).